Release binary does not work in a Docker container

I may investigate this later, but thought I should post this in case someone else has any tips:

```bash
Resolving deltas: 100% (28797/28797), done.
Error: magefile.go:6:2: could not import bytes (cannot find package "bytes" in any of:
	/go/src/github.com/gohugoio/hugo/vendor/bytes (vendor tree)
	/Users/finchnat/src/github.com/golang/go/src/bytes (from $GOROOT)
	/go/src/bytes (from $GOPATH))
```

I use the Linux binary from the v1 release.

This still doesn't work: 

```bash
Error: mage.go:5:9: could not import fmt (cannot find package "fmt" in any of:
	/Users/finchnat/src/github.com/golang/go/src/fmt (from $GOROOT)
	/go/src/fmt (from $GOPATH))
root@a7699ceb0bec:/go# mage -version
Mage Build Tool 1.0.1
Build Date: 2017-10-16T03:55:08Z
Commit: e66e3df1f2a53233ad48e5e874a0e532c65592dd
root@a7699ceb0bec:/go#
```

`mage.go` is a very simple build file and it, obviously, also fails for more complex scenarios:

```go
// +build mage

package main

import ("fmt")

func Build() {
    fmt.Println("hi")
}
```

Ahh crap, I realize now that the test I ran did not actually test the bug..
I forgot to import something in my magefile.

Gah, this seems like a compiler bug.

On Mon, Oct 16, 2017, 3:28 AM Bjørn Erik Pedersen <notifications@github.com>
wrote:

> This still doesn't work:
>
> Error: mage.go:5:9: could not import fmt (cannot find package "fmt" in any of:
> 	/Users/finchnat/src/github.com/golang/go/src/fmt (from $GOROOT)
> 	/go/src/fmt (from $GOPATH))
> root@a7699ceb0bec:/go# mage -version
> Mage Build Tool 1.0.1
> Build Date: 2017-10-16T03:55:08Z
> Commit: e66e3df1f2a53233ad48e5e874a0e532c65592dd
> root@a7699ceb0bec:/go#
>
> mage.go is a very simple build file and it, obviously, also fails for
> more complex scenarios:
>
> // +build mage
> package main
> import ("fmt")
> func Build() {
>     fmt.Println("hi")
> }
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/magefile/mage/issues/69#issuecomment-336802691>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADCcyOK8Vt7-IzNZRbV_OgHWp2C_M_Dsks5sswWugaJpZM4P5r5Y>
> .
>

--- a/mage/main.go
+++ b/mage/main.go
@@ -101,6 +101,7 @@ func ParseAndRun(dir string, stdout, stderr io.Writer, stdin io.Reader, args []s
 		log.Println(initFile, "created")
 		return 0
 	}
+
 	return Invoke(inv)
 }
 
@@ -183,6 +184,7 @@ func Invoke(inv Invocation) int {
 	for i := range files {
 		fnames = append(fnames, filepath.Base(files[i]))
 	}
+
 	info, err := parse.Package(inv.Dir, fnames)
 	if err != nil {
 		log.Println("Error:", err)
@@ -264,24 +266,8 @@ func Magefiles(dir string) ([]string, error) {
 
 // Compile uses the go tool to compile the files into an executable at path.
 func Compile(path string, stdout, stderr io.Writer, gofiles []string) error {
-	// we explicitly set GOROOT on the command because go build will fail
-	// without it.
-	goroot := os.Getenv("GOROOT")
-	if goroot == "" {
-		c := exec.Command("go", "env", "GOROOT")
-		c.Stderr = stderr
-		b, err := c.Output()
-		if err != nil {
-			return fmt.Errorf("failed to get GOROOT from 'go env': %v", err)
-		}
-		goroot = strings.TrimSpace(string(b))
-		if goroot == "" {
-			return errors.New("could not determine GOROOT")
-		}
-	}
-
 	c := exec.Command("go", append([]string{"build", "-o", path}, gofiles...)...)
-	c.Env = append(os.Environ(), "GOROOT="+goroot)
+	c.Env = os.Environ()
 	c.Stderr = stderr
 	c.Stdout = stdout
 	err := c.Run()

--- a/parse/parse.go
+++ b/parse/parse.go
@@ -3,12 +3,14 @@ package parse
 import (
 	"fmt"
 	"go/ast"
+	"go/build"
 	"go/doc"
 	"go/parser"
 	"go/token"
 	"go/types"
 	"log"
 	"os"
+	"os/exec"
 	"strings"
 )
 
@@ -105,7 +107,7 @@ func getPackage(path string, files []string, fset *token.FileSet) (*ast.Package,
 
 	pkgs, err := parser.ParseDir(fset, path, filter, parser.ParseComments)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to parse directory: %v", err)
 	}
 
 	for name, pkg := range pkgs {
@@ -117,6 +119,21 @@ func getPackage(path string, files []string, fset *token.FileSet) (*ast.Package,
 }
 
 func makeInfo(dir string, fset *token.FileSet, files map[string]*ast.File) (types.Info, error) {
+	goroot := os.Getenv("GOROOT")
+	if goroot == "" {
+		c := exec.Command("go", "env", "GOROOT")
+		b, err := c.Output()
+		if err != nil {
+			return types.Info{}, fmt.Errorf("failed to get GOROOT from 'go env': %v", err)
+		}
+		goroot = strings.TrimSpace(string(b))
+		if goroot == "" {
+			return types.Info{}, fmt.Errorf("could not determine GOROOT")
+		}
+	}
+
+	build.Default.GOROOT = goroot
+
 	cfg := types.Config{
 		Importer: getImporter(fset),
 	}
@@ -133,7 +150,10 @@ func makeInfo(dir string, fset *token.FileSet, files map[string]*ast.File) (type
 	}
 
 	_, err := cfg.Check(dir, fset, fs, &info)
-	return info, err
+	if err != nil {
+		return info, fmt.Errorf("failed to check types in directory: %v", err)
+	}
+	return info, nil
 }
 
 // errorOrVoid filters the list of functions to only those that return only an

